Source/Dafny/Resolver.cs

```diff
@@ -12800,6 +12800,10 @@ private class ClonerKeepLocalVariablesIfTypeNotSet : Cloner {
         // Check if the head of the pattern is a constructor or a variable
         if (ctors.TryGetValue(idpat.Id, out ctor)) {
           /* =[3]= */
+          if (ctor != null && idpat.Arguments == null && ctor.Formals.Count != 0) {
```

Comment:
The condition of this `if` and the next one are quite similar. Consider combining the two:

```c#
if (ctor != null && idpat.Arguments == null) {
  if (ctor.Formals.Count == 0) {
    // nullary constructor without () -- so convert it to a constructor
    idpat.MakeAConstructor();
  } 
  else {
    this.reporter.Warning(MessageSource.Resolver, pat.Tok,
    $"Variable name '{pat}' has the same name as a datatype constructor that's in scope at this location");
  }
}


```